### DISCUSSION TITLE
Can you use a select statement as the andWhere clause for a JS hook?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#4138] [Q&A] [answered]
I have code like this inside a pb_hooks file.

```
        $app.dao().db()
            .select("name", "touched", "event" )
            .from("foobars")
            .where($dbx.like("event", id ))
            .andWhere("(select touched from foobars order by touched desc limit 1)")
            .limit(100)
            .orderBy( "order ASC", "created")
            .all(all);
```

I cannot tell if this is the right way to do a subquery as the where clause. And, I'm unsure if there is a way to see the final full query so I can verify it.

Basically, I want to have a table with a field called touched which stores a last touched value (a little different than the updated field, as I'll explain). Inside a hook, I update a bunch of records and store the same touched value for each record in that batch of updates. Then, I want to query the records and only retrieve the records from the last touched event.

For example, in raw sqlite:

```

sqlite> create table hello ( name text, touched text );

sqlite> insert into hello ( name, touched ) values( 'asdasd', '12345' );
sqlite> insert into hello ( name, touched ) values( 'sdflksjfdlk', '12345' );
sqlite> insert into hello ( name, touched ) values( '309403498', '12345' );
sqlite> insert into hello ( name, touched ) values( 'asdasd', '1111' );

sqlite> .mode column
sqlite> select * from hello;
name         touched
-----------  -------
asdasd       1234   
asdasd       12345  
aasdasd      12345  
aa23123132d  12345  
asdasd       12345  
asdasd       1111   

sqlite> select * from hello where touched = (select touched from hello order by touched desc limit 1);
name         touched
-----------  -------
asdasd       12345  
aasdasd      12345  
aa23123132d  12345  
asdasd       12345  

```

In other words, I want to get only rows with the "newest" touched items.

What's the right way to do this inside a JS hook?

**Comment by ganigeorgiev** [comment, marked as answer]
If you want to use raw expressions you need to wrap them in [`$dbx.exp()`](https://pocketbase.io/docs/js-database/#dbx-expraw-optparams), eg:

```js
.andWhere($dbx.exp("touched = (your sub query)"))
```

> And, I'm unsure if there is a way to see the final full query so I can verify it.

You can explore the executed queries in the std output if you start the executable with `--dev`.

Alternatively, you can manually print it with `.build().sql()` before executing the query (aka. before `execute`, `one`, `all`), something like:

```js
const q = $app.dao().db().select(...)

console.log(q.build().sql())
```

> What's the right way to do this inside a JS hook?

The query builder approach is one way to do it. Alternatively if you don't need your query to be composable and find it "clunky", you can always execute it as raw SQL string using [`newQuery(rawQuery)`](https://pocketbase.io/docs/js-database/#execute-all) (see also [Binding parameters](https://pocketbase.io/docs/js-database/#binding-parameters)).
